### PR TITLE
05-08

### PR DESCRIPTION
**Pull request by zerlinazl**
Colab notebook for automatic segmentation. Link: https://colab.research.google.com/github/WinetraubLab/3D-segmentation/blob/05-08/3D-segmentation.ipynb

### CODE REVIEW

File: 3D-segmentation.ipynb
Comment (MyYo) [inline 3D-segmentation.ipynb:1]: I recommend changing file name to lower case 3d-segmentation.ipynb

File: import_data_from_roboflow.py
Comment (MyYo) [inline import_data_from_roboflow.py:1]: I think that once you initialized from roboflow or from image then coco_path and folder path are known.
You can just store them as local variables (no need to pass these at each call)

File: import_data_from_roboflow.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    return COCO_PATH, class_ids
+
+def get_image(image_name, parent_directory):
+    """
+    Inputs:
+        image_name: full or partial image file name.
+        parent_directory: folder that contains the image. Recursive folders are ok.
+    Returns:
+        image: the image with matching file name.
```
Comment (MyYo) [inline import_data_from_roboflow.py:162]: It seems that you are returning image path not image.
I recommend to change this function to load image content and apply clahe_normalize

File: import_data_from_roboflow.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    keyframe_indices = {0, len(frame_names)-1}
+    for ann in keyframe_filenames:
+        for i, name in enumerate(frame_names):
+            if ann in name:
+                keyframe_indices.add(i)
+
+    keyframe_indices = sorted(list(keyframe_indices))
+    return keyframe_indices
+
+def clahe_normalize(image):
```
Comment (MyYo) [inline import_data_from_roboflow.py:76]: I recommend to change to a private function _clahe_normalize

File: import_data_from_roboflow.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        for file in files:
+            if image_name in file:
+                return os.path.join(root, file)
+    
+    raise FileNotFoundError(f"No matching images for {image_name} were found under {parent_directory}.")
+
+def get_mask(path_to_image, COCO_PATH, class_id):
+    """
+    Inputs: 
+        path_to_image: full image filepath
```
Comment (MyYo) [inline import_data_from_roboflow.py:174]: change path_to_image --> image_name to match with other functions in this file

**Review by MyYo — CHANGES_REQUESTED** [review]
Thanks. I request some clarifications to help the review process

File: propagate_mask_medsam2.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                        predictor.add_new_mask(
+                            inference_state=inference_state,
+                            frame_idx=next_frame_idx,
+                            obj_id=obj_id,
+                            mask=mask.squeeze(),
+                        )
+        return video_segments_f, video_logits_f
+
+    
+    def _propagate_backward(self, images_to_segment_path, end_mask,  start_keyframe_idx, end_keyframe_idx, class_id):
```
Comment (MyYo) [inline propagate_mask_medsam2.py:82]: Change to 

def _propagate_sequential_masks

'''
This function receives a list of images (images[0], images[1], ... images[n])
And a binary mask associated with image[0] (mask_0) and return a list of masks
mask[0], mask[1], ..., mask[n]
If mask_n is defined, but not mask_0, it will start from mask_n and propagate mask_n-1 ,... mask_0

Note that mask[0] is identical to mask_0
INPUTS:
images,
mask_0=None, mask_n=None
Returns:
masks_logistic

File: propagate_mask_medsam2.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                        predictor.add_new_mask(
+                            inference_state=inference_state,
+                            frame_idx=next_frame_idx,
+                            obj_id=obj_id,
+                            mask=mask.squeeze(),
+                        )
+        return video_segments_b, video_logits_b
+
+
+    def propagate_sequence(self, images_to_segment_path, class_id, coco_path):
```
Comment (MyYo) [inline propagate_mask_medsam2.py:137]: list of images images
mask_0, mask_n
0 and n are the the key frame

return masks[0],...,mask[n]

**Review by MyYo — CHANGES_REQUESTED** [review]

File: import_data_from_roboflow.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
+from roboflow import Roboflow
+from pycocotools import mask as mask_utils
+
+import os
+import cv2
+import json
+from collections import defaultdict
+import numpy as np
+
+COCO_PATH = ""
```
[the remark this answers is not in the document]
Comment (zerlinazl) [reply import_data_from_roboflow.py]: Yes, that is a better name

File: propagate_mask_medsam2.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        """
+        # Initialize predictor
+        predictor = build_sam2_video_predictor(
+            config_file= self.config_filepath,
+            ckpt_path= self.checkpoint_filepath,
+            apply_postprocessing=True,
+            # hydra_overrides_extra=hydra_overrides_extra,
+            vos_optimized=  True,
+        )
+        return predictor
```
[the remark this answers is not in the document]
Comment (zerlinazl) [reply propagate_mask_medsam2.py:32]: Good point, self.predictor works.

File: propagate_mask_medsam2.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+    def predict_mask(self, predictor, inference_state, frame_idx, prompt_mask, reverse=False):
```
[the remark this answers is not in the document]
Comment (zerlinazl) [reply propagate_mask_medsam2.py:33]: propagate_sequence is a wrapper that calls propagate forward and reverse functions, which then call predict_mask to actually generate the prediction. I could combine the forward and reverse functions into one function with reverse=True/False flag, would that be better for readability?
